### PR TITLE
Improve Configuration#withDependencies documentation

### PR DESCRIPTION
**Pull request by jvandort**
The prior documentation suggested to use the method for mutating dependencies. Given changes to the way dependency resolution handles mutation of configurations, using withDependencies for this use case is no longer recommended. Instead, for adding dependencies to a configuration lazily, users should use addLater or addAllLater on DependencySet or DependencyConstraintSet. Further, instead of mutating dependencies with withDependencies, users should add constraints that influence the resolution of other dependencies (if they're added to the graph).

We describe this in detail in the javadoc for this method. We also update the user guide to remove the recommendation of using withDependencies to mutate dependencies and instead add snippets showing the proper way to lazily add dependencies and influence dependency version selection.

Finally, we update the build to avoid withDependencies where possible.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test bd

**Comment by jvandort** [comment]
@bot-gradle test bd

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java [hunk 1]
```diff
@@ -493,27 +495,65 @@ default void fromDependencyCollector(DependencyCollector collector) {
...
      * }
      * </pre>
      *
-     * Actions will be executed in the order provided.
+     * In some cases, using this method may still be necessary:
+     *
+     * <ul>
+     *     <li>
+     *         <strong>Adding excludes</strong>: This method may be used to lazily add excludes to a Configuration.
+     *         Adding excludes to declared dependencies should be handled with {@link ComponentMetadataHandler component metadata rules}
```
Comment (octylFractal) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java:550]: Small recommendation so it reads slightly better:
```suggestion
     *         Adding excludes to declared dependencies should be handled with {@linkplain ComponentMetadataHandler component metadata rules}
```

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java [hunk 2]
```diff
@@ -493,27 +495,65 @@ default void fromDependencyCollector(DependencyCollector collector) {
...
+     *     <li>The {@link Configuration} is consumed as a variant by another project</li>
+     *     <li>Another {@link Configuration} that extends this one is resolved, published, or consumed</li>
      * </ul>
+     * <p>
+     * In general, this method should be avoided in favor of other lazy APIs. However, in some cases where
+     * lazy APIs are not yet available, this method can be used to perform actions before the configuration
+     * is used.
+     * <p>
+     * Despite the method's name, callbacks registered on this method should not add dependencies to the
+     * configuration or mutate the dependencies already present on the configuration. Instead, use the
```
Comment (octylFractal) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java:513]: Is this supposed to read more like "Despite the method's name, it should not be used to add dependencies to the configuration or mutate the dependencies already present on the configuration"? Currently it reads like "you should use this method, but not call specific things inside of your callback".

**Review by lkasso — APPROVED** [review]
Docs LGTM

**Comment by big-guy** [comment]
@bot-gradle merge

**Pull request merged**